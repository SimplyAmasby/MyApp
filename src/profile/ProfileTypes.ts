export enum Units {
    Imperial,
    Metric
}

export interface Profile {
    weight: number,
    height: number,
    units: Units
}

export enum ReducerActionType {
    HANDLE_PICKER_CHANGE,
    HANDLE_TEXT_CHANGE,
    IMPERIAL_HEIGHT_CHANGE,
    UPDATE_STATE
}

export type ProfileReducerAction = {
    updateType: ReducerActionType,
    fieldName: string,
    payload: string | Units | Profile
}