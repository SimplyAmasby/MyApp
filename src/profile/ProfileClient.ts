import { fileSystemHelper } from "../utilities/FileSystemHelper";
import { Profile } from "./ProfileTypes";

const PROFILE_FILENAME = 'ProfileData.json';

export class ProfileClient {
    saveProfileDataAsync = async (profile: Profile) => {
        if (!profile) return;
        return await fileSystemHelper.writeFileAsync(JSON.stringify(profile), PROFILE_FILENAME);
    }

    loadProfileDataAsync = async () => {
        const fileString = await fileSystemHelper.readFileAsync(PROFILE_FILENAME);
        return (fileString ? JSON.parse(fileString) : {}) as Profile;
    }
}

export const profileClient = new ProfileClient();