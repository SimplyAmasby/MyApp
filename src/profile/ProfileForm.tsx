import { Picker } from "@react-native-picker/picker";
import React, { useCallback, useState } from "react";
import { StyleSheet, Text, TextInput, TouchableOpacity, View } from "react-native";
import { Profile, ProfileReducerAction, ReducerActionType, Units } from "./ProfileTypes";


export interface ProfileFormProps {
    profile: Profile
    dispatch: (profile: ProfileReducerAction) => void
    onSave: () => void
}

export const ProfileForm = ({ dispatch, onSave, profile }: ProfileFormProps) => {

    const getWeightLabel = (): string => {
        return `${profile.units === Units.Metric ? 'kg' : 'lb'}${profile.weight > 1 ? 's' : ''}`;
    }
    const imperialHeightComponent = useCallback(
        () => {
            let feet = Math.floor(profile.height);
            let inches = Math.round((profile.height % 1) * 12);
            return (
                <>
                    <View style={styles.field}>
                        <TextInput
                            keyboardType="numeric"
                            onChangeText={value => dispatch({
                                fieldName: "feet",
                                updateType: ReducerActionType.IMPERIAL_HEIGHT_CHANGE,
                                payload: value
                            })}
                            style={styles.inputStyle}
                            value={feet.toString() || ''}

                        />
                        <Text>ft</Text>
                    </View>
                    <View style={styles.field}>
                        <TextInput
                            keyboardType="numeric"
                            onChangeText={value => dispatch({
                                fieldName: "inches",
                                updateType: ReducerActionType.IMPERIAL_HEIGHT_CHANGE,
                                payload: value
                            })}
                            style={styles.inputStyle}
                            value={inches.toString()}
                        />
                        <Text>in</Text>
                    </View>
                </>
            )
        },
        [profile.height]
    )

    const metriclHeightComponent = useCallback(
        () => (
            <>
                <View style={styles.field}>
                    <TextInput
                        keyboardType="numeric"
                        onChangeText={value => dispatch({
                            fieldName: "height",
                            updateType: ReducerActionType.HANDLE_TEXT_CHANGE,
                            payload: value
                        })}
                        style={styles.inputStyle}
                        value={profile.height?.toString() || ''}

                    />
                    <Text>m</Text>
                </View>
            </>
        ),
        [profile.height]
    )

    return !profile ? <Text>Loading</Text> : (
        <View style={styles.container}>
            <Picker
                selectedValue={profile.units}
                onValueChange={(itemValue) => dispatch({
                    fieldName: "units",
                    updateType: ReducerActionType.HANDLE_PICKER_CHANGE,
                    payload: itemValue
                })}
            >
                <Picker.Item label="Imperial" value={Units.Imperial} />
                <Picker.Item label="Metric" value={Units.Metric} />
            </Picker>
            <Text>Weight</Text>
            <View style={styles.field}>
                <TextInput
                    keyboardType="numeric"
                    onChangeText={value => dispatch({
                        fieldName: "weight",
                        updateType: ReducerActionType.HANDLE_TEXT_CHANGE,
                        payload: value
                    })}
                    value={profile.weight?.toString() || ''}
                />
                <Text>{getWeightLabel()}</Text>
            </View>

            <Text>Height</Text>
            {profile.units === Units.Metric ? metriclHeightComponent() : imperialHeightComponent()}
            <TouchableOpacity style={styles.button} onPress={onSave}>
                <Text style={{ color: '#fff', textAlign: 'center' }}>Save</Text>
            </TouchableOpacity>
        </View>
    )
}

const styles = StyleSheet.create({
    button: {
        backgroundColor: '#228B22',
        borderRadius: 20,
        marginTop: 20,
        padding: 10
    },
    container: {
        padding: 20
    },
    field: {
        flex: 1,
        flexDirection: 'row',
        justifyContent: 'space-around'
    },
    inputStyle: {
        borderColor: 'black',
        padding: 10
    }
})
