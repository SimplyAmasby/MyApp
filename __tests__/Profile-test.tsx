/**
 * @format
 */

import 'react-native';
import React from 'react';
import { ProfileForm } from '../src/profile/ProfileForm';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { Profile, Units } from '../src/profile/ProfileTypes';

it('renders correctly', () => {

  const initialState: Profile = {
    weight: 0,
    height: 0,
    units: Units.Imperial
  };
  const tree = renderer.create(
    <ProfileForm
      dispatch={() => { }}
      onSave={() => { }}
      profile={initialState} />).toJSON();
  expect(tree).toMatchSnapshot();
});
