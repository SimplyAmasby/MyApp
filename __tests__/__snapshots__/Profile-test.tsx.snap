// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 20,
    }
  }
>
  <View>
    <RNCPicker
      items={
        Array [
          Object {
            "label": "Imperial",
            "testID": undefined,
            "textColor": undefined,
            "value": 0,
          },
          Object {
            "label": "Metric",
            "testID": undefined,
            "textColor": undefined,
            "value": 1,
          },
        ]
      }
      numberOfLines={1}
      onChange={[Function]}
      selectedIndex={0}
      style={
        Array [
          Object {
            "height": 216,
          },
          undefined,
        ]
      }
    />
  </View>
  <Text>
    Weight
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      value="0"
    />
    <Text>
      lb
    </Text>
  </View>
  <Text>
    Height
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      style={
        Object {
          "borderColor": "black",
          "padding": 10,
        }
      }
      value="0"
    />
    <Text>
      ft
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-around",
      }
    }
  >
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      style={
        Object {
          "borderColor": "black",
          "padding": 10,
        }
      }
      value="0"
    />
    <Text>
      in
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#228B22",
        "borderRadius": 20,
        "marginTop": 20,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "textAlign": "center",
        }
      }
    >
      Save
    </Text>
  </View>
</View>
`;
